ownCloud/Client/PhotoSelectionViewController.swift

```diff
@@ -0,0 +1,351 @@
+//
+//  PhotoUploadViewController.swift
+//  ownCloud
+//
+//  Created by Michael Neuwert on 24.02.2019.
+//  Copyright © 2019 ownCloud GmbH. All rights reserved.
+//
+
+import UIKit
+import Photos
+import PhotosUI
+
+private extension UICollectionView {
+	func indexPathsForElements(in rect: CGRect) -> [IndexPath] {
+		let allLayoutAttributes = collectionViewLayout.layoutAttributesForElements(in: rect)!
+		return allLayoutAttributes.map { $0.indexPath }
+	}
+}
+
+typealias PhotosSelectedCallback = ([PHAsset]) -> Void
+
+class PhotoSelectionViewController: UICollectionViewController, Themeable {
+
+	// MARK: - Constants
+	fileprivate let thumbnailSizeMultiplier: CGFloat = 0.205
+	fileprivate let verticalInset: CGFloat = 1.0
+	fileprivate let horizontalInset: CGFloat = 1.0
+	fileprivate let itemSpacing: CGFloat = 1.0
+	fileprivate let thumbnailMaxWidthPad: CGFloat = 120.0
+	fileprivate let thumbnailMaxWidthPhone: CGFloat = 80.0
+
+	// MARK: - Instance variables
+	var fetchResult: PHFetchResult<PHAsset>!
+	var assetCollection: PHAssetCollection?
+	var availableWidth: CGFloat = 0
+	var selectionCallback: PhotosSelectedCallback?
+
+	fileprivate var thumbnailSize: CGSize!
+	fileprivate var previousPreheatRect = CGRect.zero
+	fileprivate var thumbnailWidth: CGFloat = 80.0
+
+	fileprivate let imageManager = PHCachingImageManager()
+	fileprivate let layout = UICollectionViewFlowLayout()
+	fileprivate lazy var durationFormatter = DateComponentsFormatter()
+
+	// MARK: - Init / deinit
+	init() {
+		layout.scrollDirection = .vertical
+		layout.sectionInset = UIEdgeInsets(top: verticalInset, left: horizontalInset, bottom: verticalInset, right: horizontalInset)
+		layout.minimumLineSpacing = itemSpacing
+		layout.minimumInteritemSpacing = itemSpacing
+		super.init(collectionViewLayout: layout)
+	}
+
+	required init?(coder aDecoder: NSCoder) {
+		fatalError("not implemented")
+	}
+
+	deinit {
+		PHPhotoLibrary.shared().unregisterChangeObserver(self)
+	}
+
+	// MARK: UIViewController life cycles
+
+	func calculateItemSize() {
+		let totalWidth = view.bounds.inset(by: view.safeAreaInsets).width
+		let totalHeight = view.bounds.inset(by: view.safeAreaInsets).height
+
+		let width = UIDevice.current.orientation.isLandscape ? totalHeight : totalWidth
+
+		if availableWidth != width {
+			availableWidth = width
+			let maxThumbnailWidth = UIDevice.current.userInterfaceIdiom == .phone ? thumbnailMaxWidthPhone : thumbnailMaxWidthPad
```

Comment:
If possible, would be awesome to use the view's `self.traitCollection.horizontalSizeClass` here instead, to account for iPad multitasking, where the app might be used in a size closer to an iPhone display than an iPad.

Reply:
fixed